Add support for __array_function__ protocol

pentschev opened the pull request:
Resolves #4563.

- [ ] Tests added / passed
- [ ] Passes `flake8 dask`

mrocklin commented:
cc @shoyer @hameerabbasi

pentschev commented:
Ok, I'm not really sure why that's failing only on newer Python/NumPy versions. Anybody sees something obvious I've missed?

hameerabbasi commented:
NumPy 1.13 and 1.11 didn't have this protocol... So what you added was essentially dead code then.

mrocklin commented:
I'm surprised to see tests with sparse and cupy in this PR.  From my perspective this is more about having Dask Arrays respond to NumPy functions, rather than having the blocks of a Dask array respond to NumPy functions.

So I would be more interested in checks like the following:

```python
x = da.ones(...)
y = np.concatenate([x, x, x])
assert isinstance(y, da.Array)
```

Where we would also want to check things like `np.linalg.qr` (which you ran into with cupy) and any others that we know are sometimes problematic.

pentschev commented:
> NumPy 1.13 and 1.11 didn't have this protocol... So what you added was essentially dead code then.

Oh yes, that makes sense. I've been testing with upstream NumPy, that's why it passed here.

pentschev commented:
@mrocklin it's good that you mention that, I've totally forgotten about that case. But either way, are you suggesting we drop Sparse and CuPy tests? I understand it may be a little early to have them properly setup due to different version requirements, but ultimately, isn't that what we want, to have Dask working also with other arrays?

mrocklin commented:
We do want those to work, yes, but I think that that effort is somewhat orthogonal from the solution you've added here.

Or, in testing terms, you've got some great integration tests here, it would be good to see some unit tests as well that just test dask array's support of the `__array_function__` protocol with normal NumPy arrays.

pentschev commented:
Ok, so here's my suggestion, I will add the tests you suggested (which I really forgot before), and skip the tests we currently have if NumPy < 1.17, does that sound acceptable?

mrocklin commented:
Sure.  More tests are always great (assuming they're decently fast)

On Fri, Mar 8, 2019 at 11:51 AM Peter Andreas Entschev <
notifications@github.com> wrote:

> Ok, so here's my suggestion, I will add the tests you suggested (which I
> really forgot before), and skip the tests we currently have if NumPy <
> 1.17, does that sound acceptable?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/dask/dask/pull/4567#issuecomment-471054678>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AASszPhczbZ13URRnlpNqoNeYL8lLmNFks5vUr9AgaJpZM4bl0pH>
> .
>

pentschev commented:
> Sure. More tests are always great (assuming they're decently fast)

Agreed, the current ones were tweaked to be fast, I didn't profile them exactly, but all 3 tests were certainly < 1 sec.

pentschev commented:
Ah, now I realized why the tests are failing, before 1.17 (currently, upstream), we have to set the environment variable NUMPY_EXPERIMENTAL_ARRAY_FUNCTION=1.

So the question here is, can we enabled that in CI for newer versions of Python? Otherwise, I think the only alternative is to skip these tests until v1.17 is released.

mrocklin commented:
I think that enabling that is fine

On Fri, Mar 8, 2019 at 12:10 PM Peter Andreas Entschev <
notifications@github.com> wrote:

> Ah, now I realized why the tests are failing, before 1.17 (currently,
> upstream), we have to set the environment variable
> NUMPY_EXPERIMENTAL_ARRAY_FUNCTION=1.
>
> So the question here is, can we enabled that in CI for newer versions of
> Python? Otherwise, I think the only alternative is to skip these tests
> until v1.17 is released.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/dask/dask/pull/4567#issuecomment-471060807>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AASszHKr2YQbaO8EN5x5O-0oUo-bz46Dks5vUsOngaJpZM4bl0pH>
> .
>

pentschev commented:
Alright, I added some `__array_function__` unit tests, it's by no means an extensive list, but probably a good starting point for us.

Regarding the NumPy version, we need at least v1.16 (now skipped for older versions) and `NUMPY_EXPERIMENTAL_ARRAY_FUNCTION=1` variable exported, which I now added. Is it possible for us to test also with NumPy v1.16.x?

shoyer commented on dask/array/tests/test_array_function.py:
```diff
@@ -0,0 +1,81 @@
+    res_y = func(y)
+
+    assert(isinstance(res_y, da.Array))
+    assert_eq(res_y, res_x)
```
Add some tests for functions that dask doesn't bother to implement? Also it would be good to test at least one function from a submodule.

pentschev replied:
Will do.

shoyer commented on dask/array/core.py:
```diff
@@ -1002,6 +1002,22 @@ def __array__(self, dtype=None, **kwargs):
+        if da_func is func:
+            return NotImplemented
+        if not all(issubclass(t, Array) for t in types):
+            return NotImplemented
```
I think most dask functions can actually handle combinations of dask arrays and numpy arrays just fine. So you might switch this to `issubclass(t, (Array,  np.ndarray))`.

mrocklin replied:
So, this is tricky, we might not want to explicitly call out `np.ndarray` if we want to support other array types as well.  For example, in theory `np.dot(dask_array, sparse_array)` should work?

hameerabbasi replied:
If we want to be completely general, we could use `hasattr(t, ‘__array_function__’)`. But this would also include Pandas dataframes at some point I’m assuming...

shoyer replied:
Right, `__array_function__` doesn't guarantee that you implement NumPy's API like NumPy.

shoyer commented on dask/array/core.py:
```diff
@@ -1002,6 +1002,22 @@ def __array__(self, dtype=None, **kwargs):
+        if da_func is func:
+            return NotImplemented
+        if not all(issubclass(t, Array) for t in types):
+            return NotImplemented
```
The `__array_function__` protocol guarantees that it will only get called if `dask.array.Array` is in types. So if that's all you wanted to check, you could drop this entirely.

The problem is that you want to support operations involving some but not all other array types. For example, `xarray.DataArray` wraps dask arrays, not the other way around. So `dask.Array.__array_function__` should return `NotImplemented` in that case, leaving it up `xarray.DataArray.__array_function__` to define the operation.

We basically need some protocol or registration system to mark arrays as "can be coerced into dask array", e.g., either
```
sparse.Array.__dask_chunk_compatibile__ = True
```
or
```
dask.array.register_chunk_type(sparse.Array)
```

mrocklin replied:
Hrm, that's an interesting problem.  Maybe we try to turn all of the args into dask arrays with `da.asarray` and, if that works, proceed?

I'm not sure that this would produce the correct behavior for `xarray.DataArray` though.  I genuinely don't know what the right behavior is there.

hameerabbasi replied:
I actually think my idea for mixins can help with this. Specifically, we could check the specific mixins that Dask absolutely requires, I’m assuming that’d be the indexing mixin.

pentschev replied:
> Hrm, that's an interesting problem. Maybe we try to turn all of the args into dask arrays with `da.asarray` and, if that works, proceed?

I actually like this idea for its simplicity.

> If we want to be completely general, we could use `hasattr(t, ‘__array_function__’)`. But this would also include Pandas dataframes at some point I’m assuming...

I also like this idea in general, but I think then it would be best before to agree on all mixins we expect to have/support. Plus, if I understood correctly, it feels like this will require much more type checking, which may become complicated to handle.

shoyer replied:
From my perspective as an xarray maintainer, casting an xarray.DataArray into a dask.array should do one of two things:
- raise TypeError, or
- pull out .data, and cast that to a Dask array if it isn't already

We don't want xarray.DataArray inside Dask arrays, because xarray objects have intentionally different behavior from NumPy arrays, so Dask array operations would not be well defined.

hameerabbasi replied:
I agree with the `TypeError`. The whole point of the protocol was to avoid these kinds of project-specific dependencies... I'm comfortable with @mrocklin's idea of casting to a `dask.Array` provided that it performs the correct checks and actually returns a Dask array (I'm assuming this would have to be done within functions anyway), although I guess at that point we're kicking the can down the road, because similar issues arise within the constructor.

hameerabbasi replied:
Also, since XArray is a container of arrays like Dask and expects a NumPy-like API, it'll face similar issues.

shoyer replied:
A related issue is maintaining a well defined type casting hierarchy: https://www.numpy.org/neps/nep-0013-ufunc-overrides.html#type-casting-hierarchy

Relying on a generic mixin would be fine, but ideally it would be a mixin that Dask.array itself does not support. Otherwise you'll end up in situations where type casting is not commutative (e.g., `type(np.stack([a, b])) != type(np.stack([b, a]))`), which we would really like to avoid. I guess Dask could safely coerce anything that implements InMemory and NumpyDuckArray mixins?

A special attribute or method seems a little easier since that's something that can be monkey patched in if desired (vs injecting another super class, which I'm sure is possible but seems tricky). I like the idea of encouraging library authors to explicitly "opt in" to Dask compatibility but not strictly requiring it.

hameerabbasi replied:
The way I see it is there are two ways to classify things:

- Implements the NumPy API or not (let's ignore for now what APIs are implemented)
- Contains NumPy-like arrays or not.

Common examples:
- CuPy (contains-> no, implements->yes)
- Sparse (contains-> no (but could contain CuPy, maybe... A bit tricky because more than one array), implements->yes)
- Dask (contains-> yes, implements->yes)
- XArray (contains-> yes, implements->no)

We only run into the problem you're suggesting when there is more than one library where the answer to both questions is yes. And it looks like that might become the case.

So here's what I suggest: APIs that contain other arrays should have a clear protocol of extracting the underlying arrays which then _will not_ be containers.

shoyer replied:
For what it's worth I'm thinking of a similar "opt in" approach with xarray. There's no intrinsic reason for preferring `UnitsArray` wrapping `AxisLabelsArray` or the other way around, but either way it will drive our users nuts if they can't rely on the ordering of the array wrappers.

The virtue of a dask specific special method or attribute is that it allows developers of third party libraries to make an explicit decision about how they want to interact with Dask rather than making the decision for them. Refusing the temptation to guess in the face of ambiguity is the polite thing to do.

shoyer replied:
UnitsArray, e.g., as implemented in Pint, is probably a good example of another array type satisfies all of these:
1. implements NumPy's API
2. contains other duck arrays
3. could be usefully wrap or be wrapped with/by dask arrays

hameerabbasi replied:
> I guess Dask could safely coerce anything that implements `InMemory` and `NumpyDuckArray` mixins?

I was thinking of three classes of (NumPy-Specific) mixins, aimed at implementing what I see as the three major protocols:

- `NumPyUfuncMixin`
- `NumPyIndexingMixin` (for NumPy-like indexing, so `XArray` and `pandas` would skip this).
- `NumPyArrayFunctionMixin` (for implementing array methods via `__array_function__`, such as `sum` and `mean`).

I see the use of your `InMemory` mixin, but I would rename it to `InCore`, as GPU computations don't take place in main memory. :)

For now, these would be provisional, just like `__array_function__`. The hope is that these would be adopted and specific checking code would go away.

mrocklin replied:
I find the mixin idea challenging for two reasons:

1.  I think that agreeing on a few major protocols will be difficult.  Different groups will have different ideas of what these protocols are.  I think that it will be difficult to reach any sort of consensus here 
2.  I think that getting upstream projects to inherit from NumPy classes will be difficult

hameerabbasi replied:
Then how about a "numpy/pydata community repo" containing these mixins? I would be happy to maintain one. (Not in the main NumPy codebase)? The reason I like the idea of mixins is it allows "duck-array" like objects to easily implement NumPy-like functionality without too much effort.

We can also have protocols:

```
class DuckArray:
    __array_indexing__ = True
    __array_in_core__ = True
```

mrocklin replied:
I think that that would only help to address the second point above.  I'm not sure that it would solve it though.  As a worst case scenario I suspect that a project like TensorFlow would not inherit from those classes.  We could get NumPy, Dask, Sparse, and Xarray on board but I'm not sure about any of the others out there.

I personally think that duck typing is more likely to work across projects than explicit typing.

Mixins may still be a good idea for the reason you mention, you can easily implement a lot of NumPy-like functionality without too much effort, but I don't think that they're a good choice to test for Numpy-ness.

hameerabbasi replied:
I do agree that it's hard to get traction for mixins or add dependencies, but adding a protocol isn't hard.

While you're right that TensorFlow doesn't plan on implementing any of these protocols, it has always historically done its own thing, and it has its own distributed system. I suspect that with Google's historical "rewrite the world and do it better" philosophy, it'll be hard to get them to play along anyway.

In contrast, anyone specifically aiming to be a NumPy duck array won't have a problem implementing protocols, and that is what I'm planning to do, to allow mixins to implement methods and so forth based on the implementation of protocols.

pentschev replied:
For the time being, I've used @shoyer and @mrocklin suggestions replacing only that last condition in `__array_function__` to:

```python
if not all(issubclass(t, Array) for t in types):
    return NotImplemented
```

It seems to me that the mixin discussion has grown enough that we should move it to a Dask issue, discussed there and later review the current code. Anybody disagrees with my proposal?

pentschev replied:
Since nobody opposed, I moved the discussion to https://github.com/dask/dask/issues/4583.

mrocklin replied:
I think that with @shoyer 's statements below we can drop this conditional entirely.  It should always pass

>  The __array_function__ protocol guarantees that it will only get called if dask.array.Array is in types. So if that's all you wanted to check, you could drop this entirely.

hameerabbasi commented:
> Regarding the NumPy version, we need at least v1.16 (now skipped for older versions) and `NUMPY_EXPERIMENTAL_ARRAY_FUNCTION=1` variable exported, which I now added. Is it possible for us to test also with NumPy v1.16.x?

You need to add that for 1.16, where it was experimental. For 1.17, though, the env var isn't needed currently, but people are proposing that we keep the "experimental" status of the NEP for one more version...

pentschev commented on continuous_integration/travis/run_tests.sh:
```diff
@@ -13,3 +13,7 @@ else
+
+# This needs to be enabled to test __array_function__ protocol with
+# NumPy v1.16.x, enabled by default starting in v1.17
+export NUMPY_EXPERIMENTAL_ARRAY_FUNCTION=1
```
Hmm, alright, sounds like that's a better place to have this, I'll move it there.

pentschev replied:
Oh, but we don't have a build with NumPy v1.16.x, so there's nowhere to add the environment variable. It would be nice if after this commit we could start testing that, so:

1. Should I add a build for v.1.16.x (I suggest `x = 2`, the latests)?
2. Adding the build there is all we have to do, or is there some other configuration (and perhaps package installing) that needs to be done on Travis' side?

mrocklin replied:
I recommend bumping the NumPy version for the build that is currently at 1.15 to 1.16 and then adding this environment variable to that build as well.

You shouldn't have to do any other build or configuration.

pentschev commented on dask/array/tests/test_array_function.py:
```diff
@@ -0,0 +1,81 @@
+    res_x = func(x)
+    res_y = func(y)
+
+    assert(isinstance(res_y, da.Array))
```
It's ok, I'll fix that. I'm also a fan of style consistency, even though sometimes I'm not even consistent with my own style. :)

pentschev commented:
> You need to add that for 1.16, where it was experimental. For 1.17, though, the env var isn't needed currently, but people are proposing that we keep the "experimental" status of the NEP for one more version...

While I agree that features should be experimental until they can _really_ be considered stable, I was really expecting that v1.17 would introduce this feature as stable. Without that, it's difficult to even get other projects to even run unit tests for it, e.g., CuPy. I'm totally willing to help improving everything we can to attempt having this in v1.17 as stable, are there many major concerns about its maturity?

hameerabbasi commented:
> are there many major concerns about its maturity?

CuPy has been using it for a while (as short as whiles go)... So I think not. There might be features added but not removed.

hameerabbasi commented:
One thing I must mention I like about this implementation is that it specifically avoids a NumPy dependency or mapping.

mrocklin commented on dask/array/tests/test_array_function.py:
```diff
@@ -0,0 +1,100 @@
+    assert_eq(s.compute(scheduler='single-threaded'),
+              s_base.compute(scheduler='single-threaded'))
+    assert_eq(v.compute(scheduler='single-threaded'),
+              v_base.compute(scheduler='single-threaded'))
```
Note that `assert_eq` runs things single threaded anyway.  We also prefer that people not pre-compute results, that way `assert_eq` can run some additional checks on the dask object.

pentschev replied:
> Note that assert_eq runs things single threaded anyway.

What's the reasoning behind that? I particularly think this is bad, since it wouldn't test potential issues that arise from parallel execution (e.g., race conditions), or do you have tests somewhere that actually do the multi-threaded testing?

mrocklin replied:
We do have a variety of tests for each of the schedulers, yes (the dask/distributed test suite in particular is quite extensive).  However testing each of the schedulers against each of the collections (arrays, bags, delayed, dataframes, futures, ml, ...) probably isn't necessary, and tends to make debugging failures difficult. 

The separation between collections and schedulers via task graphs ends up being clean enough that single-threaded testing for the collections seems to have won out over time as the most pragmatic choice.

Regardless, even if we were to make a change, this PR probably isn't the place to make it.  We should continue to insert collections directly into assert_eq, and then if we want to change the default scheduler for all tests we can do it there in the future.

pentschev replied:
Thanks for the details, @mrocklin.

mrocklin commented on dask/array/tests/test_array_function.py:
```diff
@@ -0,0 +1,100 @@
+    y = x.map_blocks(sparse.COO)
+
+    xT = da.transpose(x).compute()
+    yT = da.transpose(y).compute()
```
Here also, it would be good to not precompute results, just check 

    assert_eq(np.transpose(x), da.transpose(x))

mrocklin commented on dask/array/tests/test_array_function.py:
```diff
@@ -0,0 +1,100 @@
+]
+
+
+@pytest.mark.parametrize('func', functions)
```
If `functions` is only being used once then I have a preference to just inline it here in the `parametrize` call rather than have a separate variable.  This removes one level of indirection for future readers of the code.  I find it nice to have values defined where they are used if they are only used in one place.  I accept that there are different views on this though.  This is just subjective.

mrocklin commented on dask/array/tests/test_array_function.py:
```diff
@@ -0,0 +1,100 @@
+    assert_eq(res_y, res_x)
+
+
+@pytest.mark.parametrize('func', unimpl_functions)
```
Same here, it would be good to have `notimpl_functions` just inlined here (also, mismatched variable name I think)

pentschev commented:
Hmm, I just noticed that the reduction tests `test_topk_argtopk*` fail due to the lack of a `sort()` implementation in Dask. This turns us back again to the discussion in https://github.com/numpy/numpy/issues/12974. While we can easily fix those (or hack those to work, depending on how one look at the problem) by simply forcing a fallback of the arrays to NumPy before calling `np.sort()`and `np.argsort()` (the results are only used as baselines), maybe it would be also interesting to have a closure on that NumPy discussion to have soon a more robust solution?

pentschev commented:
Oh, there's more in `ufunc` tests. It fails specifically for `i0`, `nan_to_num` and `sinc`, which are being tested in `test_unary_ufunc`. However, those 3 functions are defined just after the following comment in `ufunc.py`: "non-ufunc elementwise functions".

It seems to me that these 3 functions are being tested in `test_unary_ufunc` by mistake, is my assumption correct?

hameerabbasi commented:
I'm willing to author a revision to the NEP. I'd love feedback from the community!

hameerabbasi commented:
However, the proper fix would not be to "coerce" to NumPy as that'd break `sort` everywhere except NumPy arrays. What I suggest instead is to "exctract" the underlying array (call `.compute()`) and run `np.sort` on that. That should be the default implementation in Dask.

pentschev commented:
> However, the proper fix would not be to "coerce" to NumPy as that'd break `sort` everywhere except NumPy arrays. What I suggest instead is to "exctract" the underlying array (call `.compute()`) and run `np.sort` on that. That should be the default implementation in Dask.

I see what you mean about coercing. However, I'll have to disagree with you if by "default implementation" you mean Dask implementing `sort()` as a sort of wrapper to `np.sort()`, in that case, it would give the impression to users that Dask really implements it, when the following is [documented](http://docs.dask.org/en/latest/array.html):

> Operations like sort which are notoriously difficult to do in parallel, and are of somewhat diminished value on very large data (you rarely actually need a full sort). Often we include parallel-friendly alternatives like topk

Since in this PR we only need `np.sort()` for a test case, I'd rather call `.compute()` on the array in that test to simulate the behavior you suggested.

hameerabbasi commented:
Then, if the "default behaviour" inside Dask is to coerce instead of error, we'd need some kind of reverse dispatcher in general, or the `np.NotImplementedButCoercible` sentinel mentioned in the NEP. Another reason to have one. :slightly_smiling_face:

pentschev commented:
> Then, if the "default behaviour" inside Dask is to coerce instead of error, we'd need some kind of reverse dispatcher in general, or the np.NotImplementedButCoercible sentinel mentioned in the NEP. Another reason to have one. 🙂

You're absolutely right. Just for the sake of clarity, my initial thought was one of the following:

1. Automatically coerces if undocumented; or
1. Documents that there's no Dask implementation, but coerces depending on the input.

For the scope of this PR, I only wanted to avoid that we add a _deceptive_ implementation.

pentschev commented:
> Oh, there's more in `ufunc` tests. It fails specifically for `i0`, `nan_to_num` and `sinc`, which are being tested in `test_unary_ufunc`. However, those 3 functions are defined just after the following comment in `ufunc.py`: "non-ufunc elementwise functions".

I fixed this issue in my latest commit. If anyone thinks this is wrong, please let me know.

I would say this is ready for another round of review now.

jakirkham commented:
> I fixed this issue in my latest commit. If anyone thinks this is wrong, please let me know.

That looks correct to me, @pentschev. Those are not really ufuncs, but they happened to more or less work for the purposes of the original tests. So it is more correct to test them separately. Thanks for correcting these.

jakirkham commented on dask/array/tests/test_reductions.py:
```diff
@@ -473,30 +473,30 @@ def test_topk_argtopk1(npfunc, daskfunc, split_every):
     # 1-dimensional arrays
     # top 5 elements, sorted descending
-    assert_eq(npfunc(a)[-k:][::-1],
+    assert_eq(npfunc(a.compute())[-k:][::-1],
```
A small, but non-critical comment. Would it make sense to actually store the NumPy arrays above and just use them down here as opposed to calling `.compute()` repeatedly?

pentschev replied:
Yes, you're right. I'm gonna fix that.

pentschev commented:
@jakirkham thanks for the review!

jakirkham commented on dask/array/tests/test_reductions.py:
```diff
@@ -471,32 +471,35 @@ def test_topk_argtopk1(npfunc, daskfunc, split_every):
     b = da.random.random((10, 20, 30), chunks=(4, 8, 8))
 
+    npa = a.compute()
+    npb = b.compute()
```
We can certainly do this. My temptation would be to write something like this. Though I leave it up to you :)

```python
    npa = np.random.random(800)
    npb = np.random.random((10, 20, 30))

    a = da.from_array(npa, chunks=((120, 80, 100, 200, 300), ))
    b = da.from_array(npb, chunks=(4, 8, 8))
```

mrocklin commented on dask/array/tests/test_routines.py:
```diff
@@ -502,7 +502,7 @@ def test_bincount_with_weights():
     dweights = da.from_array(weights, chunks=2)
     e = da.bincount(d, weights=dweights, minlength=6)
-    assert_eq(e, np.bincount(x, weights=dweights, minlength=6))
+    assert_eq(e, np.bincount(x, weights=dweights.compute(), minlength=6))
```
Why was this change needed?

pentschev replied:
Because in this, the right argument of `assert_eq()` is used as baseline, and `x` being a NumPy array, it fails here:

```python
    @wraps(np.bincount)
    def bincount(x, weights=None, minlength=None):
        if minlength is None:
            raise TypeError("Must specify minlength argument in da.bincount")
        assert x.ndim == 1
        if weights is not None:
>           assert weights.chunks == x.chunks
E           AttributeError: 'numpy.ndarray' object has no attribute 'chunks'
```

Therefore, to me it makes sense to make both arguments NumPy arrays and let it fallback to NumPy's implementation.

hameerabbasi replied:
I think the reverse should be done, with `weights` being coerced to `da.Array` with the same chunks as `x`.

mrocklin replied:
Ah I see.  I think I agree with @pentschev here.  The left argument, `e`, is testing the dask/dask situation while the right argument is testing the numpy/numpy situation.

pentschev replied:
> I think the reverse should be done, with `weights` being coerced to `da.Array` with the same chunks as `x`.

That would cause `da.bincount()` to be dispatched, then we would be testing against `e` which computes exactly that.

pentschev commented:
> I think that with @shoyer 's statements below we can drop this conditional entirely. It should always pass

> > The array_function protocol guarantees that it will only get called if dask.array.Array is in types. So if that's all you wanted to check, you could drop this entirely.

I think I've left this for a reason, I can't remember for sure. But I think it was in the case of having arrays of different types. Wouldn't it if one array was a Dask array and another a CuPy one, for example?

mrocklin commented:
Currently the code is 

```python
        if not any(issubclass(t, (Array, np.ndarray)) for t in types):
            return NotImplemented
```

But we know that `any(issubclass(t, Array) for t in types)` is true.  If it wasn't then we wouldn't be in this function.

pentschev commented:
@mrocklin I'm confused, you're saying now we should get rid of the entire condition or just of `Array` and keep `np.ndarray`?

mrocklin commented:
I'm saying that I think we can remove the entire condition.  If what @shoyer suggests is true then I don't think it will ever be triggered.

pentschev commented:
I was thinking of other types, like CuPy and maybe Sparse. I thought I had seen one such case, but maybe I got confused. I'll remove for now, and if need be, we can always add it back later.

mrocklin commented:
Currently your condition will not trigger if any of the arguments is of `dask.array.Array` type.  

The fact that `Array.__array_function__` is triggered means that at least one of our arguments is of `dask.array.Array` type.  It doesn't matter what the other arguments are.

I may not fully understand things though.

pentschev commented:
@mrocklin I don't either, but that's most likely right, since Dask can operate on different types as well, it should probably be capable of operator on mixed types too. For now I've removed this condition.

mrocklin commented:
Merging this in.   Thanks for the effort here @pentschev and thank you @hameerabbasi, @jakirkham, and @shoyer for the review

shoyer commented:
For testing, I suppose this is OK, but I do think it's a *really* bad idea for dask not to include any type checking inside `__array_function__`. This is basically guaranteed to result in bugs.

But let's discuss more in https://github.com/dask/dask/issues/4583.

shoyer commented:
To be clear, thanks @pentschev and @mrocklin for pushing this forward. I'm really excited about the potential here!

I just want to get the design right. Given that we were involved in writing the NEP, I feel like we should try hard to be an exemplar implementation of `__array_function__`.

mrocklin commented:
Ah, my apologies if I jumped the gun here.  Happy to revert and reopen if desired.

shoyer commented:
no need to revert, we can fix this forward

pentschev commented:
Thanks for the reviews @mrocklin, @shoyer, @hameerabbasi, @jakirkham.

I absolutely agree with @shoyer, we should get the design right. I'm happy to help with whatever I can, I'm not sure I can help much now with coming up with a design since I'm not aware of many of the use cases and how they should work, but count on me later on for implementation and testing.
